### PR TITLE
Build infrastructure changes to support sflow docker and utilities

### PR DESCRIPTION
**Pull request by padmanarayana**
**- What I did**

Add build infrastructure changes to support sflow. 

**- How I did it**

  Introduce a new "sflow" container (if ENABLE_SFLOW is set). The new docker will include:
	hsflowd : host-sflow based daemon is the sFlow agent
	psample : Built from libpsample repository. Useful in debugging sampled packets/groups.
	sflowtool : Locally dump sflow samples (e.g. with a in-unit collector)

  In case of SONiC-VS, enable psample & act_sample kernel modules.

  VS' syncd needs iproute2=4.20.0-2~bpo9+1 & libcap2-bin=1:2.25-1 to support tc-sample

  tc-syncd is provided as a convenience tool for debugging (e.g. tc-syncd filter show ...)

**- How to verify it**

1. Build sonic-vs.img.gz with and without ENABLE_SFLOW set
2. With sflow enabled, check the following in vs image:
    a. act_sample/psample
    b. docker exec -it sflow /bin/bash and inspect status
    c. docker stop/start sflow
    d. Verify sflowtool, psample and tc-syncd

**- Description for the changelog**

Add build infrastructure changes to support sflow. 

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: src/sflow/hsflowd/Makefile [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+.SHELLFLAGS += -e
+
+MAIN_TARGET = $(HSFLOWD)
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% :
+	rm -fr ./host-sflow
+	git clone https://github.com/sflow/host-sflow
+
+	pushd ./host-sflow
+	git checkout -b sflow -f 08c00e2
```
Comment (lguohan) [inline src/sflow/hsflowd/Makefile]: do we have a release version?

File: files/build_templates/sonic_debian_extension.j2 [hunk 2]
```diff
@@ -300,6 +300,14 @@ sudo cp $IMAGE_CONFIGS/platform/rc.local $FILESYSTEM_ROOT/etc/
 ## copy blacklist file
 sudo cp $IMAGE_CONFIGS/platform/linux_kernel_bde.conf $FILESYSTEM_ROOT/etc/modprobe.d/
 
+# Enable psample drivers to support sFlow on vs
+{% if sonic_asic_platform == "vs" %}
+sudo tee -a $FILESYSTEM_ROOT/etc/modules-load.d/modules.conf > /dev/null <<EOF
+psample
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2]: where do we build the psample driver?

Comment (padmanarayana) [reply files/build_templates/sonic_debian_extension.j2]: https://github.com/Azure/sonic-linux-kernel/pull/94

File: src/sflow/hsflowd/Makefile (hunk 1 above)
Comment (padmanarayana) [reply src/sflow/hsflowd/Makefile]: Not yet. This will be updated once we have a release version.

File: platform/vs/docker-syncd-vs.mk [hunk 3]
```diff
@@ -12,3 +12,5 @@ $(DOCKER_SYNCD_BASE)_DBG_DEPENDS += $(SYNCD_VS_DBG) \
                                 $(LIBSAIVS_DBG)
 
 $(DOCKER_SYNCD_BASE)_RUN_OPT += -v /host/warmboot:/var/warmboot
+
+$(DOCKER_SYNCD_BASE)_BASE_IMAGE_FILES += tc-syncd:/usr/bin/tc-syncd
```
Comment (lguohan) [inline platform/vs/docker-syncd-vs.mk]: why do we need to expose tc in syncd to the base image?

File: rules/config [hunk 4]
```diff
@@ -95,3 +95,6 @@ FRR_USER_GID = 300
 
 # Default VS build memory preparation
 DEFAULT_VS_PREPARE_MEM = yes
+
+# ENABLE_SYSTEM_SFLOW - build docker-sonic-sflow for sFlow support
+ENABLE_SFLOW = n
```
Comment (lguohan) [inline rules/config]: we should make this as default enabled in the build, but in the runtime, we do not enable it.

File: rules/sflow.mk [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+SONIC_STRETCH_DEBS += $(SFLOWTOOL)
+export SFLOWTOOL
+
+# libmnl & psample packages
+
+LIBMNL_DEV_VERSION = 1.0.4-2
+export LIBMNL_DEV_VERSION
+LIBMNL_DEV = libmnl-dev_$(LIBMNL_DEV_VERSION)_amd64.deb
+$(LIBMNL_DEV)_URL = http://ftp.debian.org/debian/pool/main/libm/libmnl/libmnl-dev_1.0.4-2_amd64.deb
+SONIC_ONLINE_DEBS += $(LIBMNL_DEV)
```
Comment (lguohan) [inline rules/sflow.mk]: why not use apt-get?

File: rules/sflow.mk [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# host-sflow package
+
+HSFLOWD_VERSION = 2.0.21
+HSFLOWD_SUBVERSION = 4
+export HSFLOWD_VERSION HSFLOWD_SUBVERSION
+
+HSFLOWD = hsflowd_$(HSFLOWD_VERSION)-$(HSFLOWD_SUBVERSION)_amd64.deb
+$(HSFLOWD)_DEPENDS += $(LIBHIREDIS_DEV) $(LINUX_HEADERS) $(LINUX_HEADERS_COMMON)
+$(HSFLOWD)_SRC_PATH = $(SRC_PATH)/sflow/hsflowd
```
Comment (lguohan) [inline rules/sflow.mk]: where is the dbg package for hsflowd?

File: rules/sflow.mk [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+SONIC_MAKE_DEBS += $(HSFLOWD)
+SONIC_STRETCH_DEBS += $(HSFLOWD)
+export HSFLOWD
+
+# sflowtool package
+
+SFLOWTOOL_VERSION = 5.04
+export SFLOWTOOL_VERSION
+
+SFLOWTOOL = sflowtool_$(SFLOWTOOL_VERSION)_amd64.deb
```
Comment (lguohan) [inline rules/sflow.mk]: amd64 -> CONFIGURED_ARCH

File: rules/sflow.mk [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+SONIC_MAKE_DEBS += $(SFLOWTOOL)
+SONIC_STRETCH_DEBS += $(SFLOWTOOL)
+export SFLOWTOOL
+
+# libmnl & psample packages
+
+LIBMNL_DEV_VERSION = 1.0.4-2
+export LIBMNL_DEV_VERSION
+LIBMNL_DEV = libmnl-dev_$(LIBMNL_DEV_VERSION)_amd64.deb
+$(LIBMNL_DEV)_URL = http://ftp.debian.org/debian/pool/main/libm/libmnl/libmnl-dev_1.0.4-2_amd64.deb
```
Comment (lguohan) [inline rules/sflow.mk]: amd64 -> $(CONFIGURED_ARCH)

File: rules/sflow.mk [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+SONIC_ONLINE_DEBS += $(LIBMNL_DEV)
+
+PSAMPLE_VERSION = 1.1
+PSAMPLE_SUBVERSION = 1
+export PSAMPLE_VERSION PSAMPLE_SUBVERSION
+
+PSAMPLE = psample_$(PSAMPLE_VERSION)-$(PSAMPLE_SUBVERSION)_amd64.deb
```
Comment (lguohan) [inline rules/sflow.mk]: amd64 -> $(CONFIGURED_ARCH)

File: rules/sflow.mk (hunk 6 above)
Comment (lguohan) [reply rules/sflow.mk]: can you following the instruction here to add dbg package and src code?

https://github.com/Azure/sonic-buildimage/blob/master/README.buildsystem.md#to-enhance-debug-dockers

**Comment by padmanarayana** [comment]
sFlow docker is enabled by default. The hsflowd daemon would be started/stopped based on "config sflow enable/disable" (this is thru https://github.com/Azure/sonic-swss/commit/d45786b8ded283ed9671f5c1832b6a0903d7b792)

Revert tc-syncd. For debugging, user can issue commands from syncd docker.

sflow docker has "--pid=host". This is to facilitate CPU/mem/IO stats on the switch. Since this is a baseline feature in host-sflow, it needs to be enabled by default.

Update host-sflow to 2.0.21-6 (which has counter polling interval disable support and other fixes).

Revert dbus change since it's taken care in host-sflow repo

As per InMon sugggestion, enabled hsflowd.crash

File: platform/vs/docker-syncd-vs.mk (hunk 3 above)
Comment (padmanarayana) [reply platform/vs/docker-syncd-vs.mk]: As indicated earlier, this was added only for debugging. Now removed.

File: rules/config (hunk 4 above)
Comment (padmanarayana) [reply rules/config]: Done.

File: rules/sflow.mk (hunk 5 above)
Comment (padmanarayana) [reply rules/sflow.mk]: using apt-get

File: rules/sflow.mk (hunk 7 above)
Comment (padmanarayana) [reply rules/sflow.mk]: done

File: rules/docker-sflow.mk [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+ifeq ($(ENABLE_SFLOW), y)
+SONIC_INSTALL_DOCKER_DBG_IMAGES += $(DOCKER_SFLOW_DBG)
+SONIC_STRETCH_DBG_DOCKERS += $(DOCKER_SFLOW_DBG)
+endif
+
+$(DOCKER_SFLOW)_CONTAINER_NAME = sflow
+$(DOCKER_SFLOW)_RUN_OPT += --net=host --privileged -t
+$(DOCKER_SFLOW)_RUN_OPT += -v /etc/sonic:/etc/sonic:ro
+$(DOCKER_SFLOW)_RUN_OPT += -v /host/warmboot:/var/warmboot
+$(DOCKER_SFLOW)_RUN_OPT += --pid=host
```
[the remark this answers is not in the document]
Comment (sflow) [reply rules/docker-sflow.mk]: I believe --pid=host option makes the host /proc and /sys appear as expected.  Without this the host statistics reported by hsflowd would not be correct.  I think the alternative would be to mount them under, say, /host/proc and /host/sys and then change hsflowd to allow those new paths to be configured and used,  but even then I wouldn't be confident that everything would work the same way.

File: rules/sflow.mk (hunk 6 above)
Comment (padmanarayana) [reply rules/sflow.mk]: done

File: src/sflow/psample/Makefile [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+	git clone https://github.com/Mellanox/libpsample.git
+	cp -r debian libpsample
+
+	pushd ./libpsample
+	git checkout -b libpsample -f e48fad2
+
+	# get the dpkg lock and install libmnl-dev
+	while true; do
+	if mkdir $(DEST)/dpkg_lock &> /dev/null; then
+	{ sudo apt-get install -f -y libmnl-dev=1.0.4-2 && rm -d $(DEST)/dpkg_lock && break; } || { rm -d $(DEST)/dpkg_lock && exit 1 ; }
```
[the remark this answers is not in the document]
Comment (padmanarayana) [reply src/sflow/psample/Makefile]: done - libnml-dev is already in sonic-slave stretch docker..

File: rules/docker-sflow.mk (hunk 10 above)
[the remark this answers is not in the document]
Comment (lguohan) [reply rules/docker-sflow.mk]: there is security risk for letting process in the container to be aware of whole process. I'd like to remove this --pid if it is not absolutely needed.

[the remark this answers is not in the document]
Comment (sflow) [reply rules/docker-sflow.mk]: By "whole process" do you mean access to /proc?  There are dozens of places in the hsflowd code where information is read from /proc,  so it is needed.  Can you provide more information about the security risk that you are concerned about?

[the remark this answers is not in the document]
Comment (lguohan) [reply rules/docker-sflow.mk]: from https://docs.docker.com/engine/reference/run/

"In certain cases you want your container to share the host’s process namespace, basically allowing processes within the container to see all of the processes on the system. For example, you could build a container with debugging tools like strace or gdb, but want to use these tools when debugging processes within the container."

if you can do strace/gdb, then within the sflow container you can manipulate the host processes or processes in other container, which is posing security risk.

[the remark this answers is not in the document]
Comment (sflow) [reply rules/docker-sflow.mk]: We haven’t tested this to see if it works yet, but if we removed —pid=host and instead mounted /proc as a read-only mount point would that address your concerns?

[the remark this answers is not in the document]
Comment (lguohan) [reply rules/docker-sflow.mk]: mount /proc will let the process to access other processes in the system. it is the similiar as pid=host.

Meanwhile, if you look the hsflowd code to read cpu, i compared the /proc entries in and out of the docker. Most of the values are the same.

https://github.com/sflow/host-sflow/blob/master/src/Linux/readCpuCounters.c

[the remark this answers is not in the document]
Comment (sflow) [reply rules/docker-sflow.mk]: Maybe some /proc counters are the same as for the host as a whole,  but it's not generally true.

The hsflowd daemon is for streaming telemetry data.  It's common practice to expose /proc.  For example,  consider Google's cadvisor:

https://github.com/google/cadvisor

It runs as a docker container with these settings:

sudo docker run \
  --volume=/:/rootfs:ro \
  --volume=/var/run:/var/run:ro \
  --volume=/sys:/sys:ro \
  --volume=/var/lib/docker/:/var/lib/docker:ro \
  --volume=/dev/disk/:/dev/disk:ro \
  --publish=8080:8080 \
  --detach=true \
  --name=cadvisor \
  google/cadvisor:latest

We have already set aside container telemetry that hsflowd normally generates (which would be helpful for monitoring SONiC switches) so we are asking for significantly less access than cadvisor.

Additional examples:
https://www.elastic.co/guide/en/beats/metricbeat/6.3/running-on-docker.html
https://github.com/signalfx/docker-collectd
https://hub.docker.com/_/telegraf

Another level of constraint (which we have not tested) would be to itemize the /proc/ paths that we are allowed to read.  If we have no access to /proc/<pid>, I believe all information about other processes is shielded.  Is that the main sticking point?

The list of paths we need to read looks like this:

/proc/loadavg
/proc/stat
/proc/uptime
/proc/cpuinfo
/proc/diskstats
/proc/mounts
/proc/net/vlan/config
/proc/net/if_inet6
/proc/net/dev
/proc/net/snmp
/proc/meminfo
/proc/vmstat

If we test with only these paths exposed (read-only of course),  and we find that it works OK,  then will that address your concerns?

[the remark this answers is not in the document]
Comment (lguohan) [reply rules/docker-sflow.mk]: i am not if there are difference for these entries between the docker and host.

[the remark this answers is not in the document]
Comment (sflow) [reply rules/docker-sflow.mk]: I think you are right.  That's good.  It looks like we can drop the --pid=host.

**Pull request merged**